### ISSUE TITLE
As a user, I can customise the REST server to meet all of my requirements

### ISSUE THREAD

**Issue by sstone1** [hyperledger/composer#4094]
# Abstract
Give users of Composer the ability to generate a static REST server on disk, so that they can customise the REST server to meet all of their requirements.

# Motivation
Today, Composer includes the Composer REST server (`composer-rest-server`), an application that both generates and serves a REST API.
The application under the covers is a LoopBack 3 application that uses the Composer LoopBack connector to interact with a deployed business network.
The Composer REST server includes a bunch of code for:
- Discovering the Composer models in a business network
- Generating LoopBack models from those Composer models
- Generating APIs for those LoopBack models
- Securing those APIs with authentication (API key or Passport)
- Multi-user mode
- TLS security

At no point can the user customise any part of this code. The only route users have is to ask the Composer team to add additional command line options to the Composer REST server to support their requirements, or start from scratch and write their own REST servers.

The latter ("roll your own") is something that happens on a regular basis within IBM teams working on client projects, as well as other external companies that I have discussed the REST server with.

Over time, it has become increasingly clear that this approach is unmaintainable - users keep asking for more and more options, and we cannot keep up with the demand - nor do I think we should add a billion options to enable all of these requests. It would be easier to give the users the code so that they can customise that code to meet their requirements.

The common requests we get include:
- How can I hide certain APIs that are being generated?
- How can I customise certain APIs to add additional logic?
- How can I add my own authentication?
  - This one tends to come up when Passport isn't as pluggable as it claims, and different authentication mechanisms don't integrate nicely, e.g. `passport-local` or `passport-jwt`.
- How can I add my own APIs?

To be clear; I still believe that the Composer REST server serves a very good purpose in being able to easily generate REST APIs for a deployed business network, and that this improvement proposal would introduce an additional component alongside the Composer REST server rather than a replacement for it.

# Specification
I propose that we extend the Yeoman generator with a new generator that can generate a LoopBack 3 application for a deployed business network. This LoopBack 3 application would be equivalent in functionality to the Composer REST server, but all of the code would be generated and written out to disk for the user to edit at their discretion. For example, instead of generating LoopBack models in memory at startup, the LoopBack models would be on disk and loaded during startup.

```
$ yo
? 'Allo Simon! What would you like to do? Hyperledger Composer

Make sure you are in the directory you want to scaffold into.
This generator can also be run with: yo hyperledger-composer

Welcome to the Hyperledger Composer project generator
? Please select the type of project: LoopBack
You can run this generator using: 'yo hyperledger-composer:loopback'
Welcome to the Hyperledger Composer LoopBack project generator
? Do you want to connect to a running Business Network? Yes
? Project name: sstone1-loopback
? Description: Hyperledger Composer LoopBack project
? Author name: Simon Stone
? Author email: sstone1@uk.ibm.com
? License: Apache-2.0
? Name of the Business Network card: admin@my-network
```

This new generator would be documented under a topic of "customising the REST server", and would not form part of the developer tutorial or existing REST server documentation. This is a new feature for advanced Composer users.

# Rationale
Most of the rationale is covered by the motivation (I hope!).

I've suggested a "LoopBack" generator rather than an "REST" generator to handle a future that most likely includes an "Express" generator (along with generators for other REST API frameworks).

# Backwards Compatibility
No backwards compatibility impact; this is all additional functionality, and no changes to the Composer REST server will be done as part of the work (aside from possible internals changes to avoid code duplication).

# Test Requirements
Unit testing will be delivered for all code, as well as integration tests that ensure that a working LoopBack application is generated.

# Implementation
I intend on implementing this improvement proposal as described above. I'd love it if anyone wanted to join forces in extending this to include an "Express" generator as well as a "LoopBack" generator - having both implementations done at once would likely influence the code delivered for the better 😉 

# Copyright
This document is licensed under the Creative Commons Attribution 4.0 International Public License.

**Comment by dselman** [comment]
Sounds good. Loopback 4 is under development (and very different to Loopback 3), so you might want to include the Loopback version number in the product type.

**Comment by jverhoelen** [comment]
Thanks for this nice proposal and the good ideas that are already included.

For my current project we're also facing a somehow K.O. criterium for composer-rest-server: the fact that a REST server instance is deployed with a specific card/identity already. We would need the possibility to authenticate with the card against the REST server and the server then takes this card to authenticate against the business network/Fabric.

Do you think that something like this can also be handled by the level of customisation that your ideas may grant? Or could this be another project/module in the Composer community? How could this aspect be implemented? We'd also like to help on the implementation side as soon as the scope is clear.

**Comment by jt-nti** [comment]
Would it be useful to use the annotation functionality @dselman introduced to customise which parts of a model get generated into the rest server to start with? That could cut down on the manual customisation required after generation.

On the subject of manual customisation after generation, it's worth considering how to facilitate regeneration that avoids too many clashes with customised code as the Composer model evolves. (There should definitely be some sort of version/compatibility check on connection as Dan mentioned.)

**Comment by sstone1** [comment]
@jverhoelen I think a combination of this CIP and multi-user mode will work for your scenario, with some tweaking around the edges.

Today, the Composer REST server requires a business network card for connecting to the business network during startup and discovering the Composer models in that business network. I call this the "discovery business network card", although I'm not sure many others do.

In multi-user mode (https://hyperledger.github.io/composer/latest/integrating/enabling-multiuser), each authenticated client of the REST API must provide their own business network card to interact with the business network. These business network cards are stored in a wallet, on the Composer REST server, which is private to that authenticated user and persisted in a data store of your choice.

Additionally, in multi-user mode, the discovery business network card is never used again after startup.

This CIP would mean that the discovery takes place as a one-off by whoever is generating the REST server, and it no longer needs to take place at REST server startup or runtime. The "whoever" would have to have a business network to perform this action, but wouldn't need it afterwards.

Authenticating to the REST server using the business network card might be tricky though, I'd be interested to hear more about this requirement - and why a traditional authentication mechanism supported by Passport (like LDAP, JWT, OAuth, etc) wouldn't be suitable?

The ability to customise the REST server code as required though should mean you can implement any authentication code you want.

**Comment by sstone1** [comment]
> Would it be useful to use the annotation functionality @dselman introduced to customise which parts of a model get generated into the rest server to start with? That could cut down on the manual customisation required after generation.

:+1: I think we'll end up looking at this problem as part of a more general piece of work when we come to looking at how new Fabric capabilities such as private/local collections fit into the Composer programming model. For example, `public` and `private` might start featuring in the modelling language.

> On the subject of manual customisation after generation, it's worth considering how to facilitate regeneration that avoids too many clashes with customised code as the Composer model evolves. (There should definitely be some sort of version/compatibility check on connection as Dan mentioned.)

:+1: and a simple version number equality or semver match might be enough to satisfy this as a first pass.

**Comment by dselman** [comment]
I have a bit of an issue with model annotations for REST server generation — it basically means the model is now tied to a single network and can’t be reused easily across networks. It would be cleaner to have some network specific config imo.

**Comment by jt-nti** [comment]
@dselman I agree that there should definitely not be network specific data in the annotations. I was thinking more along the lines of `@rest("public")` or `@rest("admin")` or `@rest("etl")` to tag parts of the model and generating multiple APIs from the same model. Similar to the `public` vs. `private` annotations that @sstone1 mentioned but for filtering what gets included in the generated API.
